upgrade-2.x.sh

```diff
@@ -249,6 +250,176 @@ function device_type_match {
+            rm -rf "${sysroot:?}/"*
+            systemctl start docker-host
+            sleep 10
+        fi
+    fi
+
+    log "Starting hostapp-update"
+    hostapp-update -i "${update_package}" || log ERROR "hostapp-update has failed..."-
+}
+
+function non_hostapp_to_hostapp_update {
+    local update_package=$1
+    local sysroot="/mnt/sysroot/inactive"
+    local temp_sysroot
+
+    find_partitions
+    stop_services
+
+    # Mount spare root partition
+    umount "${update_part}" || true
+    mkfs.ext4 -F -L "${update_label}" "${update_part}"
+    temp_sysroot=$(mktemp -d)
+    mount "${update_part}" "${temp_sysroot}" || log ERROR "Cannot mount inactive partition ${update_part} to ${temp_sysroot}..."
+
+    case "${SLUG}" in
+        raspberry*)
+            log "Running pre-update fixes for ${SLUG}"
+            pre_update_pi_bootfiles_removal
+            ;;
+        *)
+            log "No device-specific pre-update fix for ${SLUG}"
+    esac
+
+    docker pull "${update_package}" || log ERROR "Couldn't pull docker image..."
+    mkfifo /tmp/resinos-image.docker
+    docker save "${update_package}" > /tmp/resinos-image.docker &
+    mkdir -p /mnt/data/resinhup/tmp
+
+    log "Starting hostapp-update"
+    # Note that the following docker daemon is started with a different --bip and --fixed-cidr
+    # setting, otherwise it is clashing with the system docker on resinOS >=2.3.0 || <2.5.1
+    # and then docker pull would not succeed
+    # The requrired volumes are:
+    #  * /mnt/sysroot/inactive: the new rootfs
+    #  * /mnt/boot: the boot partition
+    #  * /tmp/resinos-image: the FIFO file that is used to transfer the resinOS image into the container
+    #  * /mnt/data/resinhup/tmp: a location to use as docker temporary directory within the container
+    # shellcheck disable=SC2016
+    docker run \
+      --rm \
+      --name resinhup \
+      --privileged \
+      -v "${temp_sysroot}":/mnt/sysroot/inactive \
+      -v /mnt/boot:/mnt/boot \
+      -v /tmp/resinos-image.docker:/resinos-image.docker \
+      -v /mnt/data/resinhup/tmp:/mnt/data/resinhup/tmp \
+      "${update_package}" \
+      /bin/bash -c 'storage_driver=$(cat /boot/storage-driver) ; DOCKER_TMPDIR=/mnt/data/resinhup/tmp/ dockerd --log-driver=journald -s=$storage_driver --data-root='"${sysroot}"'/docker -H unix:///var/run/docker-host.sock --pidfile=/var/run/docker-host.pid --exec-root=/var/run/docker-host --bip 10.114.201.1/24 --fixed-cidr=10.114.201.128/25 --iptables=false & sleep 10; echo "Starting hostapp-update"; hostapp-update -f /resinos-image.docker' \
+    || log ERROR "Update based on hostapp-update has failed..."
+}
+
+function find_partitions {
+    # Find which partition is / and which we should write the update to
+    root_part=$(findmnt -n --raw --evaluate --output=source /)
+    log "Found root at ${root_part}..."
+    case $root_part in
+        *2)
+            # on 2.x the following device types have these kinds of results for $root_part
```

Comment:
@agherzan this is a bug. Why are not all device use label based partitions?

Reply:
I think the problem here is to detect which root we are running on. They all use labels but the problem is detecting if we are running from rootA or rootB.